### ISSUE TITLE
Should all streams raise EndOfStream on EOF?

### ISSUE THREAD

**Issue by agronholm** [agronholm/anyio#128]
Traditionally, socket-based streams have return an empty bytes object when the stream has been closed from the other end.
This is currently the case with AnyIO v2's byte streams. Object streams, on the other hand, raise `EndOfStream` because there is no other way to indicate EOF since, for bytes-oriented object streams, an empty bytes object might be a completely valid thing to send.

I'm proposing that *all* streams simply raise `EndOfStream` when the EOF has been reached. The reasoning is that code that works with both `ByteReceiveStream` and `ObjectReceiveStream[bytes]` (such as `TLSStream`) would not have to special case `ByteStream` anymore.

TL;DR – Pros:

- Unified end-of-stream handling (no more checking for an empty bytes)
- Avoid busy-waiting loops when the code keeps reading but does not check for empty bytes

Cons:

- It's a break from the tradition and might surprise users

**Comment by smurfix** [comment]
+1 from me. Returning an empty string (or zero bytes) always struck me as a stupid artefact from Unix end-of-file handling that has no place in a modern framework.

**Comment by nosklo** [comment]
Another great opportunity to fix APIs. Also the "surprise" in the con is "Oh, I got an unhandled EndOfStream" as opposed to "Oh, my program infinite-loops the CPU". +1

**Comment by agronholm** [comment]
Yep, I think this one has been decided.
